### ISSUE TITLE
ArrayIndexOutOfBoundsException in  net.sf.cglib.asm.ClassReader when using find(query).asList()

### ISSUE THREAD

**Issue by iAmcR** [MorphiaOrg/morphia#1223]
I'm getting an exception thrown for a relatively simply document in a collection while executing BasicDao's `find(query).asList();` Note: This is running on play framework 2.5. I'm using generics and encapsulated the call as: 

```
public List<T> find(Query<T> query) {
        return dao.find(query).asList();
 }
``` 

where `dao` is an instance of `BasicDAO<T, ObjectId>`

The exception stack is as follows: 

java.lang.ArrayIndexOutOfBoundsException: 26516
	at net.sf.cglib.asm.ClassReader.<init>(Unknown Source)
	at net.sf.cglib.asm.ClassReader.<init>(Unknown Source)
	at net.sf.cglib.asm.ClassReader.<init>(Unknown Source)
	at net.sf.cglib.asm.ClassReader.<init>(Unknown Source)
	at net.sf.cglib.proxy.BridgeMethodResolver.resolveAll(BridgeMethodResolver.java:61)
	at net.sf.cglib.proxy.Enhancer.emitMethods(Enhancer.java:911)
	at net.sf.cglib.proxy.Enhancer.generateClass(Enhancer.java:498)
	at net.sf.cglib.core.DefaultGeneratorStrategy.generate(DefaultGeneratorStrategy.java:25)
	at net.sf.cglib.core.AbstractClassGenerator.create(AbstractClassGenerator.java:216)
	at net.sf.cglib.proxy.Enhancer.createHelper(Enhancer.java:377)
	at net.sf.cglib.proxy.Enhancer.create(Enhancer.java:285)
	at com.thoughtworks.proxy.factory.CglibProxyFactory.createProxy(CglibProxyFactory.java:89)
	at com.thoughtworks.proxy.toys.hotswap.HotSwappingInvoker.proxy(HotSwappingInvoker.java:153)
	at org.mongodb.morphia.mapping.lazy.CGLibLazyProxyFactory.createProxy(CGLibLazyProxyFactory.java:67)
	at org.mongodb.morphia.mapping.ReferenceMapper.createOrReuseProxy(ReferenceMapper.java:92)
	at org.mongodb.morphia.mapping.ReferenceMapper.readSingle(ReferenceMapper.java:205)
	at org.mongodb.morphia.mapping.ReferenceMapper.fromDBObject(ReferenceMapper.java:44)
	at org.mongodb.morphia.mapping.Mapper.readMappedField(Mapper.java:848)
	at org.mongodb.morphia.mapping.Mapper.fromDb(Mapper.java:282)
	at org.mongodb.morphia.mapping.Mapper.fromDBObject(Mapper.java:193)
	at org.mongodb.morphia.query.MorphiaIterator.convertItem(MorphiaIterator.java:134)
	at org.mongodb.morphia.query.MorphiaIterator.processItem(MorphiaIterator.java:146)
	at org.mongodb.morphia.query.MorphiaIterator.next(MorphiaIterator.java:117)
	at org.mongodb.morphia.query.QueryImpl.asList(QueryImpl.java:147)
	at org.mongodb.morphia.query.QueryImpl.asList(QueryImpl.java:139)

The document is similar to the following: 

```
{
	"_id" : ObjectId("5a644342d677931235d38d0a"),
	"className" : "models.MyModel",
	"location" : DBRef("location", ObjectId("59f689c2cc753a1116105c3b")),
	"info" : {
		"username" : "myUsername",
		"emails" : [
			"xxx@xyz.com"
		],
		"name" : "First Last",
		"type" : "TYPEA"
	},
	"createdAt" : ISODate("2018-01-21T07:37:38.798Z")
}
```

Nothing in this document seems to may have triggered that. So I'm thinking this is a bug.

**Comment by iAmcR** [comment]
Some more context. The above issue occurred while running play in production mode and not while developing, where hot reload is on. 

A possible workaround for production mode: 
- The way I avoided the above was to use the 3.2.6 version of cglib-nodep instead of the default 2.2.2 version that morphia uses. For play (sbt), it was simply a matter of adding `"cglib" % "cglib-nodep" % "3.2.6"` to build.sbt. 
- The caveat is that this doesn't work while in development (hot reload mode). I therefore do not include the 3.2.6 library until needed for production. Apparently the class loaders do not load the correct class in this mode as seen in the stack trace: 
```
java.lang.RuntimeException: java.lang.ClassNotFoundException
	at org.objectweb.asm.ClassWriter.getCommonSuperClass(Unknown Source)
	at org.objectweb.asm.ClassWriter.a(Unknown Source)
	at org.objectweb.asm.Frame.a(Unknown Source)
	at org.objectweb.asm.Frame.a(Unknown Source)
	at org.objectweb.asm.MethodWriter.visitMaxs(Unknown Source)
	at net.sf.cglib.core.CodeEmitter.visitMaxs(CodeEmitter.java:842)
	at net.sf.cglib.core.CodeEmitter.end_method(CodeEmitter.java:138)
	at net.sf.cglib.proxy.InvocationHandlerGenerator.generate(InvocationHandlerGenerator.java:53)
	at net.sf.cglib.proxy.Enhancer.emitMethods(Enhancer.java:1215)
	at net.sf.cglib.proxy.Enhancer.generateClass(Enhancer.java:630)
	at net.sf.cglib.core.DefaultGeneratorStrategy.generate(DefaultGeneratorStrategy.java:25)
	at net.sf.cglib.core.AbstractClassGenerator.generate(AbstractClassGenerator.java:329)
	at net.sf.cglib.proxy.Enhancer.generate(Enhancer.java:492)
	at net.sf.cglib.core.AbstractClassGenerator$ClassLoaderData$3.apply(AbstractClassGenerator.java:93)
	at net.sf.cglib.core.AbstractClassGenerator$ClassLoaderData$3.apply(AbstractClassGenerator.java:91)
	at net.sf.cglib.core.internal.LoadingCache$2.call(LoadingCache.java:54)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at net.sf.cglib.core.internal.LoadingCache.createEntry(LoadingCache.java:61)
	at net.sf.cglib.core.internal.LoadingCache.get(LoadingCache.java:34)
	at net.sf.cglib.core.AbstractClassGenerator$ClassLoaderData.get(AbstractClassGenerator.java:116)
	at net.sf.cglib.core.AbstractClassGenerator.create(AbstractClassGenerator.java:291)
	at net.sf.cglib.proxy.Enhancer.createHelper(Enhancer.java:480)
	at net.sf.cglib.proxy.Enhancer.create(Enhancer.java:305)
	at com.thoughtworks.proxy.factory.CglibProxyFactory.createProxy(CglibProxyFactory.java:89)
	at com.thoughtworks.proxy.toys.hotswap.HotSwappingInvoker.proxy(HotSwappingInvoker.java:153)
	at org.mongodb.morphia.mapping.lazy.CGLibLazyProxyFactory.createProxy(CGLibLazyProxyFactory.java:67)
	at org.mongodb.morphia.mapping.ReferenceMapper.createOrReuseProxy(ReferenceMapper.java:92)
	at org.mongodb.morphia.mapping.ReferenceMapper.readSingle(ReferenceMapper.java:205)
	at org.mongodb.morphia.mapping.ReferenceMapper.fromDBObject(ReferenceMapper.java:44)
	at org.mongodb.morphia.mapping.Mapper.readMappedField(Mapper.java:848)
```
This is definitely a hacky work-around for my particular situation. I'm not sure if others are experiencing this, but I suspect it can happen giving the amount of dynamic runtime class/object management is happening these days. To the fix the above (or development mode), I could consider looking into CGLIB and extending the classes similar to here: https://github.com/cglib/cglib/issues/117 or look at ProxyToys to see if I can push the "right" class, so that the CGLIB finds it successfully. 

However, I'd love to here what the morphia community has to say about the use of CGLIB, considering it's original use case / assumptions are now a bit "dated" or "overloaded" and alternative code generation libraries exist?

**Comment by evanchooly** [comment]
cglib is used only for generating proxies for lazy loading.  if you don't use lazy loading, you won't see this problem.  I don't use play or have any project handy in which to test this so I'm not sure how much help I can be on that front.  However, `1.5.0-SNAPSHOT` has introduced an experimental `MorphiaReference` type that doesn't use cglib at all for lazy loading.  If you can give that a try, I'd love to hear your feedback about the feature.  I'm not sure it's the right approach but I'd love to get rid of proxy-based lazy loading as it only causes problems in projects like yours.

**Comment by iAmcR** [comment]
Thanks @evanchooly. I'll try the snapshot at some point. Do you have a change-log anywhere for me to assess what might happen while running 1.5.0-SNAPSHOT? 
I noticed there's also a 2.0.0-SNAPSHOT, what's the difference there?

**Comment by evanchooly** [comment]
You can find the list of closed issues for 1.5.0 here:  https://github.com/MorphiaOrg/morphia/milestone/29?closed=1

Those 2.0.0-SNAPSHOT builds are on an old branch where I was prototyping the new 2.0 mapping bits.  I'm going to create an official 2.0 branch soon and merge the contents of those two branches.  So for now my advice is to pretend those 2.0.0-SNAPSHOT bits don't exist.  Once I create the new branch, I'll start automatically publishing 2.0.0 snapshot builds on each commit.

**Comment by evanchooly** [comment]
master is on `3.2.9` now at any rate
